include/math/utils/ndop.h

```diff
@@ -0,0 +1,221 @@
+#ifndef MATH_NDOP_H_
+#define MATH_NDOP_H_
+
+#include <algorithm>
+#include <array>
+#include <limits>
+#include <numeric>
+#include <numbers>
+#include <functional>
+#include <optional>
+#include <vector>
+
+#include "math/math.h"
+
+namespace math::utils {
+
+struct DOP14_Traits
+{
+    constexpr static size_t get_num_axises() noexcept { return std::size(axises); }
+    constexpr static size_t get_num_planes() noexcept { return get_num_axises() * 2; }
+
+    constexpr size_t get_inverse_plane_index(size_t index) const noexcept {
+        assert(index < get_num_planes());
+        return index >= get_num_axises() ? index - get_num_axises() : index + get_num_axises();
+    }
+
+    constexpr static math::float3 get_axis(size_t index) noexcept {
+        assert(index < axises.size());
+        return axises[index];
+    }
+
+    constexpr static math::float3 get_plane_normal(size_t index) noexcept {
+        assert(index < get_num_planes());
+        return index >= get_num_axises() ? -axises[index - get_num_axises()] : axises[index];
+     }
+
+private:
+    constexpr static auto sqrt3 = std::numbers::inv_sqrt3_v<float>;
+    constexpr static std::array<math::float3, 7> axises {
+        math::float3{1.0f, 0.0f, 0.0f},
+        math::float3{0.0f, 1.0f, 0.0f},
+        math::float3{0.0f, 0.0f, 1.0f},
+        math::float3{sqrt3, sqrt3, sqrt3},
+        math::float3{sqrt3, sqrt3, -sqrt3},
+        math::float3{sqrt3, -sqrt3, sqrt3},
+        math::float3{-sqrt3, sqrt3, sqrt3},
+    };
+
+    //static_assert(math::approx_equal(math::len_squared(axises[i]), 1.0f));
+};
+
+//template <auto Traits>
+class DOP : DOP14_Traits {
+   public:
+    constexpr explicit DOP() noexcept = default;
+
+    template <typename Iterator, typename UnaryFn>
+    constexpr static DOP compute(
+        Iterator vertices_begin, Iterator vertices_end, UnaryFn&& fn_get_position = std::identity{}) noexcept;
+
+    std::vector<math::float3> get_points() const noexcept;
```

Comment:
Do we need it in public interface? 
Also it seems that std::array could be used instead, to save on heap allocations